File: 프로그래머스/3/43163. 단어 변환/단어 변환.java
import java.util.*;
import java.io.*;

class Solution {
    static boolean found;
    static int answer = Integer.MAX_VALUE;
    
    public int solution(String begin, String target, String[] words) {
        found = false;
        boolean[] visited = new boolean[words.length]; //words 배열 중 이미 사용한 단어
        
        find(begin.toCharArray(), target.toCharArray(), words, visited, 0);
        
        if(found == false) return 0;
        else return answer;
    }//main 메서드
    
    private static void find(char[] now, char[] target, String[] words, boolean[] visited, int count){
        
        if(Arrays.equals(now, target)){
            found = true;
            answer = Math.min(answer, count);
            return;
        }
        
        //아직 방문하지 않았으면서 변화한 단어가 한개일 경우 바꿔
        for(int i=0; i<words.length; i++){
            if(!visited[i] && isOnlyOneDifference(now, words[i])){
                visited[i] = true;
                find(words[i].toCharArray(), target, words, visited, count+1);
                visited[i] = false; // Backtracking을 위해 visited 상태를 원래대로 되돌림
            }
        }
    }//find 메서드
    
    private static boolean isOnlyOneDifference(char[] now, String word){
        char[] words = word.toCharArray();
        int diff = 0;
        
        if(now.length != words.length) return false;
        
        for(int i=0; i<now.length; i++){
            if(now[i] != words[i]) diff++;
            if(diff > 1) return false;
        }
        
        return diff == 1;
    }
}

File: 프로그래머스/3/43163. 단어 변환/단어 변환.py
from collections import deque
def solution(begin, target, words):
    answer = 0
    
    if target not in words:
        return 0
    q = deque([(begin, 0)])
    start = list(begin)
    visited = [False] * len(words)
    
    while q:
        word, count = q.popleft()
        
        if word == target:
            return count
        
        for w in words:
            cnt = 0
            for i in range(len(word)):
                if word[i] != w[i]:
                    cnt += 1

            if cnt == 1:
                q.append((w, count+1))
                
        
    
    return answer